doc/designs/functions-for-explicitly-fetched-signature-algorithms.md

```diff
@@ -0,0 +1,181 @@
+the two uses.
+
+### Making it possible to verify with an early signature
+
+There are newer verifying algorithms that need to receive the signature
+before processing the data.  This is particularly important for streaming
+functionality.  This design proposes a mechanism to accomodate this, which
+is different than the mechanism you get with `EVP_PKEY_verify()`.
+
+Public API - API Reference
+--------------------------
+
+### For signing with `EVP_SIGNATURE`
+
+#### Primitive / Limit size oneshot
+
+``` C
+int EVP_PKEY_sign_primitive_init(EVP_PKEY_CTX *pctx,
+                                 EVP_SIGNATURE *algo,
+                                 const OSSL_PARAM params[]);
+int EVP_PKEY_sign_primitive(EVP_PKEY_CTX *ctx,
+                            unsigned char *sig, size_t *siglen, size_t sigsize,
+                            const unsigned char *tbs, size_t tbslen);
+```
+
+#### Streaming
+
+``` C
+int EVP_PKEY_sign_data_init(EVP_PKEY_CTX *pctx, EVP_SIGNATURE *algo,
+                            const OSSL_PARAM params[]);
+int EVP_PKEY_sign_data_update(EVP_PKEY_CTX *ctx,
+                              const unsigned char *in, size_t inlen);
+int EVP_PKEY_sign_data_final(EVP_PKEY_CTX *ctx,
+                             unsigned char *sig, size_t *siglen, size_t sigsize);
+```
+
+### For verifying with `EVP_SIGNATURE`
+
+#### Primitive / Limit size oneshot
+
+``` C
+/* Initializers */
+int EVP_PKEY_verify_primitive_init(EVP_PKEY_CTX *pctx,
+                                   EVP_SIGNATURE *algo,
+                                   const OSSL_PARAM params[]);
+int EVP_PKEY_verify_recover_primitive_init(EVP_PKEY_CTX *pctx,
+                                           EVP_SIGNATURE *algo,
+                                           const OSSL_PARAM params[]);
+/* Oneshot functions */
+int EVP_PKEY_verify_primitive(EVP_PKEY_CTX *pctx,
+                              const unsigned char *sig, size_t siglen,
+                              const unsigned char *tbs, size_t tbslen);
+int EVP_PKEY_verify_recover_primitive(EVP_PKEY_CTX *pctx,
+                                      unsigned char *rout, size_t *routlen,
+                                      size_t routsize,
+                                      cnnst unsigned char *sig, size_t siglen);
+```
+
+#### Streaming
+
+Do note that there's no proposed streaming definition for verify recover
+functionality.
+
+``` C
+/* Initializers */
+int EVP_PKEY_verify_data_init(EVP_PKEY_CTX *pctx, EVP_SIGNATURE *algo,
+                              const OSSL_PARAM params[]);
+/* Signature setter */
+int EVP_PKEY_CTX_set_signature(EVP_PKEY_CTX *pctx,
+                               unsigned char *sig, size_t siglen,
+                               size_t sigsize);
+/* Update and final */
+int EVP_PKEY_verify_data_update(EVP_PKEY_CTX *ctx,
+                                const unsigned char *in, size_t inlen);
+int EVP_PKEY_verify_data_final(EVP_PKEY_CTX *ctx);
```

Comment:
Similarly to signature I'd propose having also EVP_PKEY_verify_data() oneshot call.